src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -430,16 +502,31 @@ Pipe ReadFromMergeTree::spreadMarkRangesAmongStreams(
     if (0 == info.sum_marks)
         return {};
 
-    size_t num_streams = requested_num_streams;
     if (num_streams > 1)
     {
         /// Reduce the number of num_streams if the data is small.
         if (info.sum_marks < num_streams * info.min_marks_for_concurrent_read && parts_with_ranges.size() < num_streams)
-            num_streams = std::max((info.sum_marks + info.min_marks_for_concurrent_read - 1) / info.min_marks_for_concurrent_read, parts_with_ranges.size());
+            num_streams = std::max(
+                (info.sum_marks + info.min_marks_for_concurrent_read - 1) / info.min_marks_for_concurrent_read, parts_with_ranges.size());
     }
 
-    return read(std::move(parts_with_ranges), column_names, ReadType::Default,
-                num_streams, info.min_marks_for_concurrent_read, info.use_uncompressed_cache);
+    return read(
+        std::move(parts_with_ranges),
+        column_names,
+        ReadType::Default,
+        num_streams,
+        info.min_marks_for_concurrent_read,
+        info.use_uncompressed_cache);
+}
+
+Pipe ReadFromMergeTree::spreadMarkRangesAmongStreams(RangesInDataParts && parts_with_ranges, const Names & column_names)
+{
+    auto read = [this, &column_names](RangesInDataParts && parts_with_ranges_, size_t num_streams_)
+    {
+        return spreadMarkRangesAmongStreamsImpl(std::move(parts_with_ranges_), column_names, num_streams_);
+    };
+    return outputPerPartitionIfRequested(
```

Comment:
I would rather avoid template function and callback here.
I mean, it's always `spreadMarkRangesAmongStreamsImpl` here. Maybe make `outputPerPartitionIfRequested` just a member function?

Also, I don't like the name `outputPerPartitionIfRequested`. Maybe something like `partitionedRead` or `readFromSeparatePartitions` or similar

Reply:
changed